### PR TITLE
feat(opentelemetry source): support trace ingestion

### PR DESCRIPTION
**Pull request by caibirdme**
This PR supports ingesting trace data in opentelemetry-protocol

Here's the example config:
```yaml
sources:
  foo:
    type: opentelemetry
    grpc:
      address: 0.0.0.0:4317
    http:
      address: 0.0.0.0:4318
      keepalive:
        max_connection_age_jitter_factor: 0.1
        max_connection_age_secs: 300
sinks:
  bar:
    type: console
    inputs: [foo.traces]
    encoding:
      codec: json
  baz:
    type: console
    inputs: [foo.logs]
    encoding:
      codec: logfmt
```

we can use [opentelemetry-trace-gen](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/cmd/telemetrygen) to test it.

1. install: `go install github.com/open-telemetry/opentelemetry-collector-contrib/cmd/telemetrygen@latest`
2. produce some demo traces: `telemetrygen traces --otlp-insecure --duration 30s`
3. produce some demo logs: `telemetrygen logs --duration 30s --otlp-insecure`

### CODE REVIEW

**Comment by gcuberes** [comment]
@caibirdme, by any chance are you also working on an opentelemetry sink for traces?

**Comment by caibirdme** [comment]
> @caibirdme, by any chance are you also working on an opentelemetry sink for traces?

@gcuberes In technical terms, the issue of converting data to the OpenTelemetry (OTel) format should not be difficult. The process only involves serializing the data into a specified Protocol Buffers (PB) format and sending it via gRPC or HTTP to a fixed endpoint. However, there are some challenges in terms of how to use this approach.

Firstly, users can modify the received trace data through the VRL, which makes it difficult to ensure that the final sent data still conforms to the OTel protocol specifications (some fields may be missing or have incorrect types). Currently, I have not found any method to constrain the schema of sink data.

Secondly, if we want to convert any logs into OTel traces, users need to configure all the required fields for OTel traces, which can be cumbersome in terms of usage.

If the goal is simply to receive OpenTelemetry traces and then forward them unchanged to another location using the OTel protocol, similar to a proxy, this may be an acceptable solution for some use cases but I don't know if this is what people want.

In summary, while technically feasible, there are some challenges in terms of ensuring data conformance to the OTel protocol and ease of use for the end-users.

**Comment by caibirdme** [comment]
Anybody take a look on this pr?

**Comment by StephenWakely** [comment]
Hi, sorry this one slipped my attention. Vector has pretty poor support for traces, the only sink that can send out traces is the Datadog sink, but even then the format of those traces needs to be pretty precise. Anything else would be fairly error prone. I'm curious what your use case for this source would be?

**Comment by caibirdme** [comment]
We're using clickhouse to store the logs&traces, so clickhouse sink is enough for us.  The path looks like `client -> send otlp trace/log -> vector -> vrl -> clickhouse`. And we want to try loki in the future, the loki sink is also supported

**Comment by jszwedko** [comment]
> We're using clickhouse to store the logs&traces, so clickhouse sink is enough for us. The path looks like `client -> send otlp trace/log -> vector -> vrl -> clickhouse`. And we want to try loki in the future, the loki sink is also supported

Gotcha. So you are converting the trace to a log event to send it to one of those sinks? I'm ok with us adding this, it's just that the user experience will be pretty rough so I'd want to label it as experimental in the docs. Do you mind making the docs updates by adding a new "how it works" section here: https://github.com/vectordotdev/vector/blob/58a4a2ef52e606c0f9b9fa975cf114b661300584/website/cue/reference/components/sources/opentelemetry.cue#L197-L207 ?

**Comment by caibirdme** [comment]
> Gotcha. So you are converting the trace to a log event to send it to one of those sinks? I'm ok with us adding this, it's just that the user experience will be pretty rough so I'd want to label it as experimental in the docs. Do you mind making the docs updates by adding a new "how it works" section here:

Sure, but I don't know exactly what should be to put into `how it works`.  You mean how the source ingests trace data and converts it into log event?

**Comment by jszwedko** [comment]
> > Gotcha. So you are converting the trace to a log event to send it to one of those sinks? I'm ok with us adding this, it's just that the user experience will be pretty rough so I'd want to label it as experimental in the docs. Do you mind making the docs updates by adding a new "how it works" section here:
> 
> Sure, but I don't know exactly what should be to put into `how it works`. You mean how the source ingests trace data and converts it into log event?

Right, I would just add something like about it being experimental, have limited processing functionality available, and that the internal data format may change in the future.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Nice. The code is looking pretty solid, just a fairly minor suggestion. Can you add an integration tests that covers this [here](https://github.com/vectordotdev/vector/blob/master/src/sources/opentelemetry/integration_tests.rs).

This will also need a changelog entry. Can you add one [here](https://github.com/vectordotdev/vector/tree/master/changelog.d).

File: src/sources/opentelemetry/integration_tests.rs [hunk 1]
```diff
@@ -97,6 +104,81 @@ async fn receive_logs_legacy_namespace() {
...
+            http: HttpConfig {
+                address: source_http_address().parse().unwrap(),
+                tls: Default::default(),
+                keepalive: Default::default(),
+            },
+            acknowledgements: Default::default(),
+            log_namespace: Default::default(),
+        };
+
+        let (sender, trace_output, _) = new_source(EventStatus::Delivered);
```
Comment (StephenWakely) [inline src/sources/opentelemetry/integration_tests.rs:156]: This function will capture the `LOGS` stream from the source. This integration test is interested in the `TRACES` output. You will either want to pass the output we want to capture as a parameter to this function, or create a new `new_traces_source` function.

**Review by StephenWakely — CHANGES_REQUESTED** [review]

**Review by aliciascott — APPROVED** [review]
No docs review required

**Comment by caibirdme** [comment]
@StephenWakely pls look at this

**Comment by StephenWakely** [comment]
Not sure why that changelog job is erroring. I'll get someone to take a look..

**Comment by StephenWakely** [comment]
Not sure why that changelog job is failing. I'll ask someone to take a look.

File: changelog.d/opentelemetry_trace.feature.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Support ingesting opentelemetry traces(Experimentally)
+
+authors: caibirdme
+
```
Comment (StephenWakely) [inline changelog.d/opentelemetry_trace.feature.md:4]:
```suggestion
```

The CI job is unable to find the authors because it is looking in the last line of the file for it.

**Review by StephenWakely — CHANGES_REQUESTED** [review]

File: website/cue/reference/components/sources/opentelemetry.cue [hunk 3]
```diff
@@ -203,6 +209,13 @@ components: sources: opentelemetry: {
 				[OpenSSL configuration file](\(urls.openssl_conf)). The file location defaults to
 				`/usr/local/ssl/openssl.cnf` or can be specified with the `OPENSSL_CONF` environment variable.
 				"""
+		},
+		traces: {
+			title: "Ingest OTLP traces"
+			body: """
+				Vector now support ingest opentelemetry trace in proto format(JSON is not supported). The trace
+				data is then converted as log, and you can use vrl to remap it to whatever you want.
```
Comment (jszwedko) [inline website/cue/reference/components/sources/opentelemetry.cue:217]:
```suggestion
				Vector now support ingest opentelemetry trace in proto format(JSON is not supported). The trace
				data is then converted as log, and you can use vrl to remap it to whatever you want.
```

I don't think this is true, is it? It is a trace event (granted it shares the same internal structure as a log). I would suggest just dropping this since we don't have a "how it works" section for the log data type.

Comment (jszwedko) [reply website/cue/reference/components/sources/opentelemetry.cue:217]:
```suggestion
				Trace support is experimental and subject to change as Vector has no strongly-typed structure for traces internally. Instead traces are stored as a key/value map similar to logs. This may change in the future to be a structured format.
```

Or I'm remembering now that we wanted to explicitly call this out as experimental. Could we change it to something like this?

Comment (jszwedko) [reply website/cue/reference/components/sources/opentelemetry.cue:217]: I think we should add a note up under `notices` here too that says `Ingestion of traces is experimental`.

**Comment by caibirdme** [comment]
the failed checks seem like not related to this PR, any idea? @StephenWakely @jszwedko

**Comment by jszwedko** [comment]
> the failed checks seem like not related to this PR, any idea? @StephenWakely @jszwedko

Ah, yes, apologies if you merge in `master` it should fix that CI issue.

File: website/cue/reference/components/sources/opentelemetry.cue [hunk 4]
```diff
@@ -203,6 +209,12 @@ components: sources: opentelemetry: {
 				[OpenSSL configuration file](\(urls.openssl_conf)). The file location defaults to
 				`/usr/local/ssl/openssl.cnf` or can be specified with the `OPENSSL_CONF` environment variable.
 				"""
+		},
+		traces: {
+			title: "Ingest OTLP traces"
+			body: """
+				Trace support is experimental and subject to change as Vector has no strongly-typed structure for traces internally. Instead traces are stored as a key/value map similar to logs. This may change in the future to be a structured format.
+			"""
```
Comment (StephenWakely) [inline website/cue/reference/components/sources/opentelemetry.cue:217]:
```suggestion
				"""
```

An extra tab here should be enough to fix the cue CI error.

**Review by StephenWakely — APPROVED** [review]
Nice. Thanks!

**Comment by StephenWakely** [comment]
These regression tests should really not keep failing...

**Comment by caibirdme** [comment]
It's weird... My code should only affect opentelemetry-source

**Comment by StephenWakely** [comment]
This is now failing with doc tests:

```
failures:

---- target/debug/build/opentelemetry-proto-d2ae8386e81e6113/out/opentelemetry.proto.trace.v1.rs - proto::trace::v1::Span::attributes (line 123) stdout ----
error: expected one of `.`, `;`, `?`, `}`, or an operator, found `:`
 --> target/debug/build/opentelemetry-proto-d2ae8386e81e6113/out/opentelemetry.proto.trace.v1.rs:124:19
  |
3 | "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
  |                   ^ expected one of `.`, `;`, `?`, `}`, or an operator

error: aborting due to 1 previous error
```

On this doc comment in the generated proto file:

```rust
    /// attributes is a collection of key/value pairs. Note, global attributes
    /// like server name can be set using the resource API. Examples of attributes:
    ///
    ///      "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
    ///      "/http/server_latency": 300
    ///      "example.com/my_attribute": true
    ///      "example.com/score": 10.239
    ///
    /// The OpenTelemetry API specification further restricts the allowed value types:
    /// <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute>
    /// Attribute keys MUST be unique (it is not allowed to have more than one
    /// attribute with the same key).
    #[prost(message, repeated, tag = "9")]
```

I'm not sure I understand why..

**Review by StephenWakely — CHANGES_REQUESTED** [review]
The doc tests are failing because the comments [here](https://github.com/vectordotdev/vector/blob/master/lib/opentelemetry-proto/src/proto/opentelemetry-proto/opentelemetry/proto/trace/v1/trace.proto#L173-L176) are indented by 4 spaces, so the doc tester thinks that it is Rust code in the comments. This PR didn't change that code, but I'm guessing that since nothing used it prior to this PR the code wasn't pulled in.

It's annoying to have to do, but I think the only way to fix this would be to update those lines in the comments so they only indent by 2 spaces.

Would you be able to make this change?

**Review by StephenWakely — APPROVED** [review]
